Macroize riscv

leekillough opened the pull request.

devinamatthews committed:
Merge tlb- and slab/rr-specific gemm macrokernels. (#711)

Details:
- Merged the tlb-specific gemm macrokernel (_var2b) with the slab/rr-
  specific one (var2) so that a single function can be compiled with
  either tlb or slab/rr support, depending on the value of the
  BLIS_ENABLE_JRIR_TLB, _SLAB, and _RR. This is done by incorporating
  information from both approaches: the start/end/inc for the JR and IR
  loops from slab or rr partitioning; and the number of assigned
  microtiles, plus the starting IR dimension offset for all iterations
  after the first (ir_next). With these changes, slab, rr, and tlb can
  all be parameterized by initializing a similar set of variables prior
  to the jr loop.
- Removed the wrap-around logic that sets the "b_next" field of the
  auxinfo_t struct, which executes during the last IR iteration of the
  last JR iteration. The potential benefit of this code is so minor
  (and hinges on the microkernel making use of the b_next field) that
  it's arguably not worth including. The code also does the wrong
  thing for some threads whenever JR_NT > 1, since only thread 0 (in the
  JR group) would even compute with the first micropanel of B.
- Re-expressed the definition of bli_is_last_iter_slrr so that slab and
  tlb use the same code rather than rr and tlb.
- Adjusted the initialization of the gemm control tree accordingly.

leekillough committed:
Use here-document for 'configure --help' output. (#714)

Details:
- Changed the configure script function that outputs "--help" text to do 
  so via so-called "here-document" syntax for improved readability and
  maintainability. The change eliminates hundreds of echo statements and 
  makes it easier to change existing configure options' help text, along
  with other benefits such as eliminating the need to escape double-
  quote characters (").

leekillough committed:
Typecast printf() args to avoid compiler warnings. (#716)

Details:
- In bli_thread_range_tlb.c, typecast integer arguments passed to 
  printf() -- which are typically disabled unless debugging -- to type 
  "long" to guarantee a match to the "%ld" format specifiers used in 
  those calls. This avoids spurious warnings with certain compilers in 
  certain toolchain environments, such as 32-bit RISC-V (rv32iv).

leekillough committed:
rename files

leekillough committed:
Macroize assembly kernels so that single- and double-precision routines use the
same code, and real and complex kernels use common source code for saving and
restoring registers.

.inc file extensions are used instead of .S file extensions, because .S file
extensions cause the BLIS Makefile to create associated .o targets.

Declare assembly language kernels as taking intptr_t parameters, which are
integers the same size as pointers, and add static assertions, handled by a
new macro called bli_static_assert(), to make sure that sizeof(dim_t) <=
sizeof(intptr_t) and sizeof(inc_t) <= sizeof(intptr_t). We do not want dim_t
and/or inc_t to be declared as 64-bit on RISC-V 32-bit.

Add assert( rs_c == 1 ) to make sure an assumption of the assembly code is met.

leekillough committed:
delete file prior to renaming

leekillough committed:
renaming

leekillough committed:
add functions

leekillough committed:
remove column major assumption

leekillough commented:
x

angsch commented on kernels/rviv/3/bli_czgemm_rviv_asm_4vx4.inc:
```diff
@@ -0,0 +1,768 @@
+    // Arbitrarily long length
+    // TODO: Find a better way of querying MAXVL
+    li tmp, 256
+    vsetvli s0, tmp, VTYPE, m1, ta, ma
```
```suggestion
    vsetvli s0, zero, VTYPE, m1, ta, ma
```
The workaround with `tmp` can be skipped

leekillough replied:
I am moving all of the calculations to C so that we can use `get_vlenb()` and do not need to express the length in words, just bytes.

angsch replied:
I am not sure if I understand this comment. `vsetvli` defines how the vector data is interpreted. I don't think it is possible to get rid of this.

leekillough replied:
Yes, there is that aspect of setting vector type and other things. I was asking in the context of the MAXVL determination, since previously we had a hack which used the instruction for two purposes.

I need to add `k==0` check, set the vector mode, and a few other things.

angsch commented on kernels/rviv/3/bli_czgemm_rviv_asm_4vx4.inc:
```diff
@@ -0,0 +1,768 @@
+
+    // Arbitrarily long length
+    // TODO: Find a better way of querying MAXVL
+    li tmp, 256
```
```suggestion

```

angsch commented on kernels/rviv/3/bli_czgemm_rviv_asm_4vx4.inc:
```diff
@@ -0,0 +1,768 @@
+    slli s0, s0, DATASHIFT      // vector length in bytes
+
+    // Workaround for `mul rs_c, rs_c, s0` because rv64iv does not support `mul`
+    mv rs_c, s0                 // Assumes unit row skip
```
There is an inconsistency here. This command overwrites `rs_c` and, hence, renders the caller's value `rs_c * num_fp32_per_vector()` superfluous.

leekillough replied:
I am moving all of the calculations to the C code, which makes it simpler. The strides we pass to the assembly are measured in bytes and do not need to be multiplied or shifted. The `DATASHIFT` macro is eliminated.

angsch replied:
:+1:

angsch commented on kernels/rviv/3/bli_rviv_utils.h:
```diff
@@ -34,4 +34,32 @@
+inline uint32_t num_fp32_per_vector(void)
+{
+    uint32_t velem = 0;
+    __asm__ volatile
+    (
+    " li a5, 256                         \n\t"
+    " vsetvli a4, a5, e32, m1, ta, ma    \n\t"
+    " mv %[velem], a4                    \n\t"
+    : [velem] "=r" (velem)
+    :
+    : "a4", "a5"
+    );
```
```suggestion
inline uint32_t get_vlenb(void)
{
    uint32_t vlenb = 0;
    asm volatile (
       " csrr %0, vlenb    \n\t"    // vector length in bytes
      : "=r" (vlenb)
    ); 
    return vlenb;
}
inline uint32_t num_fp32_per_vector(void)
{
    return get_vlenb() / sizeof(float);
}
inline uint32_t num_fp64_per_vector(void)
{
   return get_vlenb() / sizeof(double);
}
inline uint32_t num_scmplx_per_vector(void)
{
   return get_vlenb() / sizeof(scomplex);
}
inline uint32_t num_dcmplx_per_vector(void)
{
   return get_vlenb() / sizeof(dcomplex);
}
```
There is one drawback with this solution: csrr is an atomic read, so likely not that fast

leekillough replied:
Do we really need four functions for `float`, `double`, `scomplex` and `dcomplex` ?  VLEN is guaranteed to be a power of 2 so if we divide by 4, 8, or 16 and then multiply back the size, we will have an identity operation and can use VLEN directly instead of dividing it and multiplying it back.  By moving all of the multiplication / shifts out of the assembly code into the C, the multiplications are much simpler and the division cancels out.

Why is this better than `vsetvli vlen, zero, ...`?  By atomic do you mean A extension is required or simply that `csrr` is atomic?

angsch replied:
> Do we really need four functions for float, double, scomplex and dcomplex

I understood your comment that you wanted to have versions for scomplex and dcomplex. If you prefer to multiply or divide the output to adjust for the data type, sure. That is what is done in the context file.

csrr appears to be the only instruction by which vlenb can be read off. My understanding is csrr is not part of A, but simply an atomic instruction that is designed to read machine data. For vlenb I do not see why it is necessary to have this as an atomic read. I assume that this is inherited behavior from reading out other machine data where atomic read and write behavior is needed.

leekillough replied:
If you see my latest changes, I use `get_vlenb()` directly to compute byte strides. I only divide it in the context routine, where it already explains why `float` is used.

angsch commented on kernels/rviv/3/bli_cgemm_rviv_4vx4.c:
```diff
@@ -54,23 +55,27 @@ void bli_cgemm_rviv_4vx4
        scomplex*  restrict a,
        scomplex*  restrict b,
        scomplex*  restrict beta,
-       scomplex*  restrict c, inc_t rs_c0, inc_t cs_c0,
```
The reference kernels in BLIS have all three parameters on one single line.

angsch commented on kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:
```diff
@@ -0,0 +1,396 @@
+    addi B_row_ptr, B_row_ptr, 8*DATASIZE
+
+    // Load A(:,l+1)
+    VLE A01, (A0_ptr)
```
nit: Not need to have a macro here, `vle.v A01, (A0_ptr)` picks up the right version of vle (vle32.v, vle64.v)

leekillough replied:
```
kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:161: Error: unrecognized opcode `vle.v v30,(t1)'

kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:302: Error: unrecognized opcode `vse.v v9,(t3)'
```

angsch replied:
Interesting...another compiler problem

leekillough replied:
I suggest keeping the macros for now.

angsch commented on kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:
```diff
@@ -0,0 +1,396 @@
+    .text
```
just noticed that the copyright header is missing in the assembly files

leekillough committed:
Add copyright messages, make date 2023

Use csrr instruction to get vlenb

Use vlenb directly in byte stride calculations
(avoids mutiplying and dividing by word size)

Remove DATASHIFT macro

Reformat C matrix arguments to match reference kernels

leekillough committed:
Merge remote-tracking branch 'flame/master' into macroize_riscv

leekillough commented:
One more thing we need to do:

We need to see if there's a way to manually add `Makefile` dependencies between the `.o` files and the `.inc` files which are `#include`d from `.S` files. We don't want to use `.S` filenames for the `.inc` include files, because then the BLIS build system will create `.o` targets for them.

Right now if you `touch` the `.inc` files, it does not cause the `.o` files to be rebuilt.

@devinamatthews @fgvanzee

leekillough committed:
add vsetvli

leekillough committed:
Rename .inc files to .h files to handle build dependencies since changes to .h
files already cause rebuilds

Add k == 0 handling

Replace fa4 = 0 hack with more descriptive name fzero which is zero throughout
entire routine

leekillough commented:
> One more thing we need to do:
> 
> We need to see if there's a way to manually add `Makefile` dependencies between the `.o` files and the `.inc` files which are `#include`d from `.S` files. We don't want to use `.S` filenames for the `.inc` include files, because then the BLIS build system will create `.o` targets for them.
> 
> Right now if you `touch` the `.inc` files, it does not cause the `.o` files to be rebuilt.
> 
> @devinamatthews @fgvanzee

I renamed the `.inc` files to `.h` files. That will "apply the hammer" if there are any changes.

leekillough committed:
Move instructions to hide latency better

angsch commented:
thank you
